src/Domain/Structure/VertexIterator.hpp

```diff
@@ -0,0 +1,31 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+
+#include "Domain/Structure/Side.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+
+/*!
+ * \brief Iterate over vertices in a `Dim`-dimensional hypercube
+ */
```

Comment:
Thanks for reviewing @moxcodes! Here's some "pseudocode" that illustrates how I currently use these iterators in the Schwarz weighting functions:

```cpp
// Computing weights on subdomains that overlap across corners and edges
for (VertexIterator<2> corner{}; corner; ++corner) {
  // Skip corners to external boundaries
  if (skip_corner(corner)) {
    continue;
  }
  for (size_t d = 0; d < 2; ++d) {
    get(corner_weight) *= intruding_weight(
        logical_coords.get(d), overlap_widths[d], corner.side(d));
  }
  // ...
}
for (EdgeIterator<2> edge{}; edge; ++edge) {
  const size_t d = edge.dimension();
  const Side side = edge.side(0);
  // Skip edges to external boundaries
  if (skip_edge(edge)) {
    continue;
  }
  get(edge_weight) *= intruding_weight(
      logical_coords.get(d), overlap_widths[d], side);
  // ...
}
```

These iterators are similar to the `SliceIterator` and `StripeIterator`. They encode the current state in the iterator object, which can be passed around (e.g. to the `skip_corner` and `skip_edge` functions).

Do you think these hypercube iterators can be useful, or should I inline their logic where I iterate over corners or edges in the Schwarz code? I'm not very attached to this code, but thought it could be a nice re-usable way of abstracting away the hypercube geometry. I'm happy to hear your suggestions!

Reply:
I do think that this sort of iterator can be useful -- from the code snippet, the only improvement that I'd suggest is try to construct the iterator in such a fashion that it could be used with range-based iteration (perhaps with some type of begin/end factory functions for the iterator?).

The edge part of this code seems counter-intuitive to me -- it has the appearance that if I have a consistent way of ordering the sides outside of the edge iterator, the edge iterator means different things as it advances (because it computes the upper/lower via a bitset of length dim-1) -- so as the iterator advances, if I'm interested in a particular side index, I will need to decrement it or omit the query to `edge.side` depending on what `edge.dimension` is.
Would it make the interface more difficult for your use-case if the `edge.side` function took the full range of dimensions and returned a `boost::optional`, giving `boost::none` if the side requested is invalid for the current edge?